Fixes #8987 - DataExporter: OutputFormat doesn't render correctly on exports

jasonex7 opened the pull request:
Fixes #8987

jasonex7 committed:
Fixes #8987 - DataExporter: OutputFormat doesn't render correctly on exports

melloware commented:
I think you can eliminate most of this PR.  You don't need to do this `RenderKit` stuff if you simply want to write to a temporary writer we do this using our `FastStringWriter` all over in the code base.

```java
ResponseWriter writer = context.getResponseWriter();

try (FastStringWriter fsw = new FastStringWriter()) {
                fsw.write("{");
                writeOption(fsw, "@context", model.getContext(), false);
                writeOption(fsw, "@type", model.getType(), true);
                encodeItems(fsw, model.getItemListType(), model.getItems(), true);
                fsw.write("}");

                writer.write(fsw.toString());
}
```

melloware commented on primefaces/src/main/java/org/primefaces/renderkit/RendererUtils.java:
```diff
@@ -70,4 +77,46 @@ else if (partialSelected) {
+        }
+
+        return ((RenderKitFactory) FactoryFinder.getFactory(FactoryFinder.RENDER_KIT_FACTORY)).getRenderKit(context, renderKitId);
+    }
```
Ithink you can remove this method

jasonex7 replied:
OK, I removed to favor the use of `cloneWithWriter` method

melloware commented on primefaces/src/main/java/org/primefaces/util/ComponentUtils.java:
```diff
@@ -607,4 +619,43 @@ public static String getDynamicColumnValueAsString(UIComponent component) {
+     */
+    public static String encodeHtml(UIComponent component, FacesContext context) {
+        ResponseWriter originalWriter = context.getResponseWriter();
+        StringWriter output = new StringWriter();
```
Just use our FastStringWriter.

jasonex7 replied:
FastStringWriter is not a javax.faces.context.ResponseWriter so I cannot pass directly to `context.setResponseWriter` method. I think that is the reason why the method needs the render kit instance, to create a ResponseWriter.

I've just commited a modifed version using FastStringWriter instead of StringWriter

jasonex7 replied:
I used `cloneWithWriter` method to avoid the use of getRenderKit method

jasonex7 commented:
> I think you can eliminate most of this PR. You don't need to do this `RenderKit` stuff if you simply want to write to a temporary writer we do this using our `FastStringWriter` all over in the code base.
> 
> ```java
> ResponseWriter writer = context.getResponseWriter();
> 
> try (FastStringWriter fsw = new FastStringWriter()) {
>                 fsw.write("{");
>                 writeOption(fsw, "@context", model.getContext(), false);
>                 writeOption(fsw, "@type", model.getType(), true);
>                 encodeItems(fsw, model.getItemListType(), model.getItems(), true);
>                 fsw.write("}");
> 
>                 writer.write(fsw.toString());
> }
> ```

The most uses I see of StringWriter is in render classes to generate strings which at the end are going to be writed to `ResponseWriter` of FacesContext:

```java
public static void encodeModel(FacesContext context, JsonLDModel model) throws IOException {
    if (model != null) {
        ResponseWriter writer = context.getResponseWriter();

        try (FastStringWriter fsw = new FastStringWriter()) {
            fsw.write("{");
            writeOption(fsw, "@context", model.getContext(), false);
            writeOption(fsw, "@type", model.getType(), true);
            encodeItems(fsw, model.getItemListType(), model.getItems(), true);
            fsw.write("}");

            writer.write(fsw.toString());
        }
    }
}
```

jasonex7 committed:
Fixes #8987 - Use FastStringWriter instead of StringWriter

melloware commented:
Hmmm but isn't that all you are doing here capturing the string of encodeAll to put in the Excel cell?

melloware commented:
If you search the code base for "setResponseWriter" you will see we do this in other places in the code

jasonex7 committed:
Fixes #8987 - Use cloned ResponseWriter

jasonex7 commented:
OK, I used a cloned writer as in TimelineRenderer and UICacheRenderer

melloware commented:
Excellent that is what I meant!

jasonex7 committed:
Fixes #8987 - Adjust javadoc

melloware commented on primefaces/src/main/java/org/primefaces/util/ComponentUtils.java:
```diff
@@ -99,7 +103,12 @@ else if (submittedValue != null) {
             if (value == UNDEFINED_VALUE) {
-                value = valueHolder.getValue();
+                if (valueHolder instanceof HtmlOutputFormat) {
+                    value = encodeHtml((HtmlOutputFormat) valueHolder, context);
```
You don't need to pass `(HtmlOutputFormat)valueHolder` here just pass `component` since its already a UIComponent that the signature of encodeHtml needs.

melloware commented on primefaces/src/main/java/org/primefaces/util/ComponentUtils.java:
```diff
@@ -607,4 +616,40 @@ public static String getDynamicColumnValueAsString(UIComponent component) {
+     * @throws UncheckedIOException Whenever something fails at I/O level. This would be quite unexpected as it happens locally.
+     * @see <a href="https://github.com/omnifaces/omnifaces">Omnifaces</a>
+     */
+    public static String encodeHtml(UIComponent component, FacesContext context) {
```
I would just call this `encodeComponent` instead of `encodeHtml` and check the Javadoc above to make sure its all still true?  Especially all the OmniFaces stuff.

jasonex7 replied:
Please check now

jasonex7 committed:
Fixes #8987 - Change method name to encodeComponent and adjust javadoc

jasonex7 committed:
Fixes #8987 - Remove unnecessary cast

jasonex7 commented on primefaces/src/main/java/org/primefaces/util/ComponentUtils.java:
```diff
@@ -607,4 +616,29 @@ public static String getDynamicColumnValueAsString(UIComponent component) {
+    public static String encodeComponent(UIComponent component, FacesContext context) {
+        ResponseWriter originalWriter = context.getResponseWriter();
+        FastStringWriter output = new FastStringWriter();
+        context.setResponseWriter(originalWriter.cloneWithWriter(output));
```
@melloware this line throws NPE because originalWriter return null. I think that is why Omnifaces uses getRenderKit method to address that situations. Maybe is because the render response is not available yet when this method is executed

jasonex7 committed:
Fixes #8987 - Consider when current responseWriter is null

Maybe if this method is called in apply request phase when responseWriter was not already set (e.g. with immediate="true").

melloware commented:
Ahhh I see I bet its not set yet because DataExporter is a TagHandler not a Renderer.
